### PR TITLE
Clean Backoff column family from wrong jobs

### LINKED ISSUE #15954: Job backoff migration causes duplicate incidents
**Describe the bug**

With https://github.com/camunda/zeebe/issues/14329 we fixed an issue that would cause failed jobs with `retries > 0` to not be re-activated after a migration, i.e. the jobs aren't placed in the `JOB_BACKOFF` column family. 

However, the fix doesn't check if a failed job has `retries > 0` or if it already has an incident, which wouldn't make it activatable.

The above leads to a job with an incident being made activatable. When this type of job is processed, and failed again, it raises a second incident. If Zeebe has the `ZEEBE_BROKER_EXPERIMENTAL_CONSISTENCYCHECKS_ENABLEPRECONDITIONS` env var set to `true`, then a consistency violation error will be thrown by RocksDB, crashing the partition.

**To Reproduce**

1. Use Zeebe `8.1.21`, `8.2.19`, `8.3.4`, or `8.4.0`.
2. Set the `ZEEBE_BROKER_EXPERIMENTAL_CONSISTENCYCHECKS_ENABLEPRECONDITIONS` env var to `true`.
3. Set the `retries` of a job to `0` and `fail` it.
4. Migrate to the next patch version of Zeebe (`8.1.22`, `8.2.20`, `8.3.5`)
5. Activate the same job (from [3]) and fail it again.
6. An error will be thrown that an entry already exists in the `INCIDENT_JOBS` column family, and the partition will crash.

**Expected behavior**

Jobs with incidents are not made activatable after a migration.

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
10:18:09.846 [Broker-0] [zb-actors-1] [StreamProcessor-1] ERROR io.camunda.zeebe.logstreams - Actor StreamProcessor-1 failed in phase STARTED.
io.camunda.zeebe.db.ZeebeDbInconsistentException: Key DbForeignKey[inner=DbLong{2251799819040844}, columnFamily=JOBS, match=Full, skip=io.camunda.zeebe.db.impl.DbForeignKey$$Lambda/0x00007f751880b920@57433d1d] in ColumnFamily INCIDENT_JOBS already exists
	at io.camunda.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.assertKeyDoesNotExist(TransactionalColumnFamily.java:297) ~[classes/:?]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.lambda$insert$0(TransactionalColumnFamily.java:81) ~[classes/:?]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.lambda$ensureInOpenTransaction$19(TransactionalColumnFamily.java:325) ~[classes/:?]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.DefaultTransactionContext.runInTransaction(DefaultTransactionContext.java:31) ~[classes/:?]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.ensureInOpenTransaction(TransactionalColumnFamily.java:324) ~[classes/:?]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.insert(TransactionalColumnFamily.java:76) ~[classes/:?]
	at io.camunda.zeebe.engine.state.instance.DbIncidentState.createIncident(DbIncidentState.java:85) ~[classes/:?]
	at io.camunda.zeebe.engine.state.appliers.IncidentCreatedApplier.applyState(IncidentCreatedApplier.java:30) ~[classes/:?]
	at io.camunda.zeebe.engine.state.appliers.IncidentCreatedApplier.applyState(IncidentCreatedApplier.java:17) ~[classes/:?]
	at io.camunda.zeebe.engine.state.appliers.EventAppliers.applyState(EventAppliers.java:335) ~[classes/:?]
	at io.camunda.zeebe.engine.processing.streamprocessor.writers.ResultBuilderBackedEventApplyingStateWriter.appendFollowUpEvent(ResultBuilderBackedEventApplyingStateWriter.java:56) ~[classes/:?]
	at io.camunda.zeebe.engine.processing.streamprocessor.writers.ResultBuilderBackedEventApplyingStateWriter.appendFollowUpEvent(ResultBuilderBackedEventApplyingStateWriter.java:42) ~[classes/:?]
	at io.camunda.zeebe.engine.processing.job.JobFailProcessor.raiseIncident(JobFailProcessor.java:173) ~[classes/:?]
	at io.camunda.zeebe.engine.processing.job.JobFailProcessor.failJob(JobFailProcessor.java:135) ~[classes/:?]
	at io.camunda.zeebe.engine.processing.job.JobFailProcessor.lambda$processRecord$0(JobFailProcessor.java:86) ~[classes/:?]
	at io.camunda.zeebe.util.Either$Right.ifRightOrLeft(Either.java:381) ~[classes/:?]
	at io.camunda.zeebe.engine.processing.job.JobFailProcessor.processRecord(JobFailProcessor.java:85) ~[classes/:?]
	at io.camunda.zeebe.engine.Engine.process(Engine.java:127) ~[classes/:?]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.batchProcessing(ProcessingStateMachine.java:356) ~[classes/:?]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.lambda$processCommand$2(ProcessingStateMachine.java:272) ~[classes/:?]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.ZeebeTransaction.run(ZeebeTransaction.java:98) ~[classes/:?]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.processCommand(ProcessingStateMachine.java:272) ~[classes/:?]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.tryToReadNextRecord(ProcessingStateMachine.java:231) ~[classes/:?]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.readNextRecord(ProcessingStateMachine.java:207) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:94) ~[classes/:?]
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:47) [classes/:?]
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:120) [classes/:?]
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:130) [classes/:?]
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:108) [classes/:?]
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:227) [classes/:?]
10:18:09.857 [Broker-0] [zb-actors-0] [ZeebePartition-1] ERROR io.camunda.zeebe.broker.system - StreamProcessor-1 failed, marking it as dead: StreamProcessor-1{status=DEAD, issue=HealthIssue[message=null, throwable=io.camunda.zeebe.db.ZeebeDbInconsistentException: Key DbForeignKey[inner=DbLong{2251799819040844}, columnFamily=JOBS, match=Full, skip=io.camunda.zeebe.db.impl.DbForeignKey$$Lambda/0x00007f751880b920@57433d1d] in ColumnFamily INCIDENT_JOBS already exists, cause=null]}
```

</p>
</details>

**Environment:**
- OS:  Camudna SaaS and SM
- Zeebe Version:  `8.1.22`, `8.2.20`, `8.3.5`, `8.4.0`
- Configuration:  `ZEEBE_BROKER_EXPERIMENTAL_CONSISTENCYCHECKS_ENABLEPRECONDITIONS=true`

**Workaround**
* Setting `ZEEBE_BROKER_EXPERIMENTAL_CONSISTENCYCHECKS_ENABLEPRECONDITIONS=false` will avoid the issue from happening until a fix is provided.

**Additional context:**
* The bug can also be reproduced by starting a `StandaloneBroker` (on the `stable/8.3` branch) with [this (internal) partition data](https://drive.google.com/file/d/19SDLOKRu6jwxDeG05CNHrIL1wYZYL5ft/view?usp=sharing).
* The problematic code is [here](https://github.com/camunda/zeebe/blob/86db453f1aef24e2fb02e7a739dda12a12b1997c/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java#L288) (on branch `main`/`8.4`) since we don't exclude jobs with incidents/`retries < 1`.
* [Support case](https://jira.camunda.com/browse/SUPPORT-20055)
* [Support case](https://jira.camunda.com/browse/SUPPORT-20113)

### PR DESCRIPTION
**Pull request by nicpuppa**
## Description

Due to https://github.com/camunda/zeebe/issues/14329 a new migration was introduced to restore the backoff column family with the correct data.

Perhaps, we missed to cover some cases (see https://github.com/camunda/zeebe/issues/15954).

 With https://github.com/camunda/zeebe/pull/16047 a fix of the migration was added, but the wrong data were not cleaned up

 This PR add a method to clean up the wrong data in the backoff column family

## Related issues

closes #15954 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by nicpuppa** [comment]
> @nicpuppa this needs backporting down to 8.1, right?

I think so

File: engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java
Comment (koevskinikola) [inline engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java:553]: :wrench: Maybe we can make the method name more descriptive. Currently, it's not clear what a "wrong" job is. How about `removeJobsWithoutRetries()`?

File: engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java
Comment (koevskinikola) [inline engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java:555]: :wrench: I think we can improve the variable and method names here. 

The method `getBackoffJobKey()` returns a `Set` of keys. It would be better if we rename it to `getBackoffJobKeys()` instead, so it is clear that multiple keys are returned.

The variable name is `backoffJobs` but it only contains the **keys** of those jobs. It would be more readable if we name it `backoffKeys` or `backoffJobKeys`. That way we will be consistent with the naming in the line above. It will also be less confusing when we read line 557 below since it looks like we remove jobs by keys there.

File: engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java
Comment (koevskinikola) [inline engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java:557]: :wrench: I'm not sure that we need to do this since `backoffJobs` is always a subset of `failedJobs`, i.e. any job added in the `backoffColumnFamily` must be in state `FAILED`.

@nicpuppa are you doing this in case a job that is not `FAILED` ended up in the `backoffColumnFamily`?

File: engine/src/test/java/io/camunda/zeebe/engine/state/migration/JobBackoffRestoreMigrationTest.java
Comment (koevskinikola) [inline engine/src/test/java/io/camunda/zeebe/engine/state/migration/JobBackoffRestoreMigrationTest.java:185]: :wrench: It would be good if we assert that the correct job is kept.

**Review by koevskinikola — CHANGES_REQUESTED** [review]
@nicpuppa the PR looks good. I have some suggestions for improving the readability of the code as well as the test coverage.

File: engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java
Comment (nicpuppa) [reply engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java:557]: > @nicpuppa are you doing this in case a job that is not `FAILED` ended up in the `backoffColumnFamily`?

It should not happen, so I will remove the retain

**Comment by nicpuppa** [comment]
Thanks for the review @koevskinikola, I applied all the suggestions. Please have another look 👀

**Review by koevskinikola — APPROVED** [review]
LGTM @nicpuppa :+1:

**Comment by koevskinikola** [comment]
@nicpuppa there seem to be some merge conflicts, can you resolve them before merging the  PR?

@Zelldon would you like to review the PR before merging?

**Comment by Zelldon** [comment]
Sorry I have limited capacity currently, but I try my best to review it

**Review by ChrisKujawa — COMMENTED** [review]
I'm bit a confused right now. Could we check what cases we would now have and what this means now.

Like: we have jobs which are activiable but shouldn't (retries 0), we have jobs in backoff which shouldn't? @nicpuppa could you think about all the cases again and write them down so we can verify that we don't miss again something. I feel otherwise we might miss again something (which I would like to avoid)

**Comment by nicpuppa** [comment]
> I'm bit a confused right now. Could we check what cases we would now have and what this means now.

All the cases that we have are:
- Failed job (retries > 0) are not in the backoff column (covered by migration)
- Jobs activable with retries = 0 are in the backoff column family (this PR should fix this)

Those should be all the cases.

cc @koevskinikola @Zelldon

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java
Comment (nicpuppa) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java:286]: In this way we can get all the failed jobs not present in the backoff column and restore it

**Comment by megglos** [comment]
@nicpuppa what's the state here for merging, feels unfortunate that this fix wasn't included in the current patch round, do you await approval by @Zelldon ?

**Comment by nicpuppa** [comment]
> @nicpuppa what's the state here for merging, feels unfortunate that this fix wasn't included in the current patch round, do you await approval by @Zelldon ?

Not really, but if @Zelldon have time to review this PR again would be good.

**Comment by Zelldon** [comment]
@nicpuppa @megglos Unfortunately I don't have enough capacity to give here more guidance, but I think we should evaluate more whether we miss any points, as I mentioned [earlier](https://github.com/camunda/zeebe/pull/16508#pullrequestreview-1901280662). 

As this is already the third PR fixing this related issue, I feel we might need to think more carefully about what else we have to check. It feels to get more risky. As we now might get into issues with different patch versions, and different migration states. If one has applied all patches, one none, one only single patch, does this make any issues during the next patch and migration, etc?

**Comment by megglos** [comment]
@oleschoenburg given you have context of the original backoff issue https://github.com/camunda/zeebe/issues/14329 and worked on rolling updates, could you help out here with a review and if needed sync with @nicpuppa ? I would like to close this case but given @Zelldon comment I would appreciate if you can assist here, while your rolling update focus that could be valuable here is fresh :)

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java
Comment (oleschoenburg) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java:281]: :thought_balloon: I first wanted to make sure I understand what exactly happens in this method. I did this by inlining all helper methods and then doing a bit of refactoring without changing behavior. This is the result:

```java
  @Override
  public void restoreBackoff() {

    final var jobsWithBackoff = new HashSet<Long>();
    backoffColumnFamily.forEach(
        (key, value) -> jobsWithBackoff.add(key.second().inner().getValue()));

    jobsWithBackoff.forEach(
        key -> {
          jobKey.wrapLong(key);
          final var jobRecord = jobsColumnFamily.get(jobKey);
          if (jobRecord != null && jobRecord.getRecord().getRetries() <= 0) {
            final var backoff = jobRecord.getRecord().getRecurringTime();
            if (backoff > 0) {
              jobKey.wrapLong(key);
              backoffKey.wrapLong(backoff);
              backoffColumnFamily.deleteIfExists(backoffJobKey);
            }
          }
        });

    final var jobsWithBackoffAfterCleanup = new HashSet<Long>();
    backoffColumnFamily.forEach(
        (key, value) -> jobsWithBackoffAfterCleanup.add(key.second().inner().getValue()));

    final var failedJobKeys = new HashSet<Long>();
    statesJobColumnFamily.forEach(
        (key, value) -> {
          if ((State.FAILED).equals(value.getState())) {
            failedJobKeys.add(key.inner().getValue());
          }
        });
    failedJobKeys.removeAll(jobsWithBackoffAfterCleanup);
    failedJobKeys.forEach(
        key -> {
          jobKey.wrapLong(key);
          final var jobRecord = jobsColumnFamily.get(jobKey);
          if (jobRecord != null
              && jobRecord.getRecord().getRecurringTime() > -1
              && jobRecord.getRecord().getRetries() > 0) {
            final var backoff = jobRecord.getRecord().getRecurringTime();
            if (backoff > 0) {
              jobKey.wrapLong(key);
              backoffKey.wrapLong(backoff);
              backoffColumnFamily.insert(backoffJobKey, DbNil.INSTANCE);
            }
          }
        });
```

So it first iterates over all backoff entries and looks up the corresponding job record. If there is one, no retries remain and the job requested a backoff, we remove a matching backoff entry (not necessarily the one we're looking at in the current iteration, that one could have a different backoff, at least in theory).

Then we go over all failed jobs that don't already have a matching backoff entry and lookup the job record. If the job has remaining retries and requested a backoff, we create a new backoff entry.

There are a couple of things that stand out to me:
1. We are populating two in-memory hashsets with all job keys that occur as backoff entries _twice_.
2. We ignore backoff entries where no matching job record can be found. Such entries should be deleted instead.
3. We only delete backoff entries that match exactly to the job record. What happens to backoff entries for the same job key but with a different backoff?

With this in mind, I've tried to simplify the implementation further and came up with the following:
```java
  @Override
  public void restoreBackoff() {
    final var jobsWithBackoff = new LongHashSet();
    backoffColumnFamily.forEach(
        (key, value) -> {
          final var jobRecord = jobsColumnFamily.get(jobKey);
          if (jobRecord == null || jobRecord.getRecord().getRetries() <= 0) {
            backoffColumnFamily.deleteExisting(key);
          } else {
            jobsWithBackoff.add(jobKey.getValue());
          }
        });

    statesJobColumnFamily.forEach(
        value -> {
          if (!State.FAILED.equals(value.getState())) {
            return;
          }
          if (jobsWithBackoff.contains(jobKey.getValue())) {
            return;
          }
          final var jobRecord = jobsColumnFamily.get(jobKey);
          final var backoff = jobRecord.getRecord().getRecurringTime();
          final var retries = jobRecord.getRecord().getRetries();
          if (backoff > 0 && retries > 0) {
            backoffKey.wrapLong(backoff);
            backoffColumnFamily.insert(backoffJobKey, DbNil.INSTANCE);
          }
        });
  }
```
This feels much simpler and more reliable:
1. all backoff entries that don't have a job or where the job has no remaining retries get deleted.
2. all failed jobs that don't already have a backoff entry, have a remaining retry and have requested a backoff create a new backoff entry.

Not only is this more efficient because we only iterate over both column families once and only keep one in-memory set of jobs with backoff. To me this is also much easier to understand.

:thought_balloon: In the future, I think we could drastically simplify this further. What we are doing is essentially starting the backoff column family new from scratch and creating necessary entries by iterating through all failed jobs that should have a backoff. This would be too inefficient to do every time but since 8.5 we only run migrations on version changes and we could further extend that so that the migration only runs if the previous version was affected by relevant bugs and skip otherwise.

**Review by lenaschoenburg — APPROVED** [review]
I've spent some time to understand what exactly happens during the migration and now feel confident enough that this behaves correctly, even when skipping in-between patches.

:wrench: However, I feel like the implementation is both difficult to understand and not very efficient. Please consider the refactoring I suggested below.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java
Comment (nicpuppa) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java:281]: Thanks @oleschoenburg for the suggestions, they looks great ❤️ I will push the updates soon 🙏

**Comment by oleschoenburg** [comment]
@nicpuppa we could also consider merging this into 8.5.0 still. Might be risky because we have no time to test it but it's the same for the patch releases anyway :shrug:

**Comment by nicpuppa** [comment]
> @nicpuppa we could also consider merging this into 8.5.0 still. Might be risky because we have no time to test it but it's the same for the patch releases anyway 🤷

I would like to ping @megglos on this

**Comment by megglos** [comment]
> > @nicpuppa we could also consider merging this into 8.5.0 still. Might be risky because we have no time to test it but it's the same for the patch releases anyway 🤷
> 
> I would like to ping @megglos on this

I agree with Ole let's Backport this to 8.5 as well, given we include it in all patches too.

**Pull request merged**
